doc/designs/quic-design/connection-state-machine.md

```diff
@@ -0,0 +1,916 @@
+arbitrary BIO to an SSL object; not only that, separate BIOs can be passed for
+the read and write directions. The nature of this BIO can be arbitrary; it could
+be a socket, or a memory buffer.
+
+Implementation of QUIC requires handling of timer events as well as the
+circumstances where a network socket becomes readable or writable. In many cases
+we need to handle these events simultaneously (e.g. wait until a socket becomes
+readable, or a timeout expires, whichever comes first).
+
+### Use of non-blocking I/O
+
+These requirements make it more or less a requirement that we use non-blocking
+network I/O; we need to be able to (at a minimum) have timeouts on recv() calls,
+and make best effort (non blocking) send() and recv() calls.
+
+The only sensible way to do this portably is to configure the socket into
+non-blocking mode. We could try to do a select() before calling send() or recv()
+to get a guarantee that the call will not block, but this will probably run into
+issues with buggy OSes which generate spurious socket readiness events. In any
+case, relying on this to work reliably is not advisable.
+
+Timeouts could be handled via setsockopt() socket timeout options, but this
+depends on OS support and adds another syscall to every network I/O operation.
+It also has obvious thread safety concerns if we want to move to concurrent use
+of a single socket at some later date.
+
+Some OSes support a `MSG_DONTWAIT` flag which allows a single I/O option to be
+made non-blocking. However some OSes (e.g. Windows) do not support this, so we
+cannot rely on this.
+
+As such, we need to configure any socket FD we use into non-blocking mode. This
+may confound users who pass a blocking socket to libssl. For most users however,
+it would be unusual for a user to pass an FD to us, then also try to send and
+receive traffic on the same socket. (There are some circumstances where this may
+occur, however; the fixed bit in the QUIC packet header may allow protocol
+multiplexing, for example with other protocols like RTP/RTCP or STUN. A
+dedicated analysis to the requirements for supporting such applications should
+probably be done to support this after MVP. In general, it is unlikely that such
+applications will use blocking I/O anyway.)
+
+As such, the impact on applications of this behaviour is limited and can be
+documented.
+
+Note that this is orthogonal to whether we provide blocking I/O semantics to the
+application. While we use non-blocking I/O internally, this can be used to
+provide either blocking or non-blocking semantics to the application, based on
+what the application requests.
+
+### Support of arbitrary BIOs
+
+However, we need to support not just socket FDs but arbitrary BIOs as the basis
+for the use of QUIC. Support for using QUIC with e.g. BIO_dgram_pair, a
+bidirectional memory buffer with datagram semantics, is to be supported as part
+of MVP. This must be rectified with the desire to support application-managed
```

Comment:
Support ..,.., is to be supported. Maybe better `Support...,..., is to be provided...`